### ISSUE TITLE
Suggestion for new hyperparameter: Regularize tree depth

### ISSUE THREAD

**Issue by mgross** [dmlc/xgboost#11845]
Currently, xgboost uses a hard cap on the tree-depth set by the "max_depth" parameter and essentially treats every tree-depth level the same. It came to my mind that a softer cap + preferring lower tree-depths might be more appropriate in some cases. 

### Motivation
The general intuition is that a-priori one could assume that first-order and lower-order feature interaction effects are generally higher in absolute terms and contain more information than higher-order feature interactions/tree depths. This would be in line with the usual testing and fitting strategy in the ANOVA literature, where one first adds the first order or main effects, then first-order interactions and so on.

### Proposed Change
Specifically, the proposal would be to shrink or decay the learning rate within an individual tree by tree depth. If the corresponding parameter (say "decay") is e.g. .8, one would shrink the learning rate (or shrink the delta step) by a factor of .8. If the learning rate is .1, it would be .08 for all splits at tree depth 2, .064 for all splits at tree depth 3 and so on.

One might argue that one could achieve a similar thing with adjusting some regularization parameter such as "lambda", but that doesn't seem to be the case; the suggestion delivers very promising results. 

### Experimental Evidence
In particular, I quickly implemented the change using the rpart package in R (using less than 10 loc) and implemented a boosting algorithm. I don't have the computational means at hand to do extensive hyperparameter optimization with inner and outer cross-validation rounds, but I used a simpler scheme to assess the effectiveness of the proposal: I chose 10 regression datasets from TabArena (https://arxiv.org/pdf/2506.16791; https://github.com/autogluon/tabarena) that employ RMSE as evaluation criterion. For each of the 10 data sets 10 train-test splits were done (2/3 train, 1/3 test). For xgboost*, n_trees, max_depth and lambda were optimized on the following grid:
```r
hp_grid_trees <- c(20, 50, 75, 100, 125, 150, 200, 300, 400, 600, 800, 1000, 1250, 1500, 2000, 2500, 3000, 4000, 5000)
hp_tree_depth <- 4:12
hp_lambda <- c(0.01, 0.1, 0.5, 1, 2, 4, 8)
```
For the proposal** the following parameter grid was chosen:
```r
hp_grid_trees <- c(20, 50, 75, 100, 125, 150, 200, 300, 400, 600, 800, 1000, 1250, 1500, 2000, 2500, 3000, 4000, 5000)
hp_decay <- c(0.95, 0.9, 0.85, 0.8, 0.75, 0.7, 0.65, 0.6, 0.55, 0.5)
```

To prove the effectivness and for computational difference tuning was done differently: For xgboost, for each train/test split, the best solution on the grid was chosen. For the proposal, the optimization was done once on a 0th train/test split and the best parameter set was then used in all 10 train/test splits. 
Clearly, this gives xgboost a significant advantage (and wouldn't be best practice as it leads to overfitting) together with an additional hyperparameter. However, the proposed method still beats classic xgboost in 9 out of 10 datasets (average RMSE):

|dataset | xgboost_best  | gbt with new hyperparameter |
|-------------| ------------- | ------------- |
|wine_quality| 0.621  | 0.614  |
|superconductivity| 9.468  | 9.391  |
|miami_housing| 84656  | 82243  |
|houses| 0.2187  | 0.2127  |
|healthcare_insurance_expenses| 4489.8  | 4489.5  |
|diamonds| 566.66  | 528.2  |
|concrete_compressive_strength| 4.145  | 4.084  |
|airfoil_self_noise| 1.471  | 1.454  |
|QSAR_fish_toxicity| 0.855  | 0.878  |
|Another-Dataset-on-used-Fiat-500| 741.1  | 735.9  |

### Drawbacks: 
Higher computational and memory needs as max_depth should be set to at least 12 (or better 14/15) and mostly more trees are needed (though this might be somewhat offset by a higher learning rate).

### Settings/Environments
* xgboost package version 1.7.11.1 in R; learning rate = 0.1,  tree_method = "exact", objective = "reg:squarederror"
** rpart package version 4.1.23 in R; learning rate = 0.1, maxdepth = 15, minsplit and minbucket = 1, cp = 0
Without using the new hyperparameter, both approaches are comparable as they yield very similar evaluation results

**Comment by trivialfis** [comment]
Hi, thank you for raising the issue and sharing your thoughts.

Would it be possible to achieve similar effects by using existing regularization techniques that penalize deep trees, like `min_split_loss`, `min_child_weight`?

**Comment by mgross** [comment]
These parameters penalize deep trees and can have a somewhat similar effect. I played a bit with them as well and it didn't seem to have the same effect as the suggestion. Also, the optimal decay was always between .75 and .5 in my evaluations meaning that even lower tree depths are already penalized pretty strong. I can do some additional evaluations, but this will take some days.

**Comment by mgross** [comment]
@trivialfis:

I ran two additional experiments: 

1. HP grid search settings as above, but with `min_child_weight` set to 5, the max value proposed in TabArena (https://arxiv.org/pdf/2506.16791). Similarly, for the adjusted rpart gbt algorithm `minsplit` and `minbucket` were also set to 5:

2. HP grid search settings as above, but instead of `lambda` (set to 1 as this is the default), `min_split_loss` is optimized as hyperparameter. As the optimal value for `min_split_loss`  is highly dependent on the scale of the target, potential values were chosen adaptively: First the scale of the target feature was measured by taking the max of IQR and MAD. Then 7 values evenly spread on log-scale from 0.02 * scale to 3*scale were used. Additionally, zero was also included giving a total of 8 potential 'min_split_loss' values per dataset.



Results for both are essentially still the same, with 9/10 cases having better evaluation error: 

|dataset | xgboost_best  | gbt with new hyperparameter |
|-------------| ------------- | ------------- |
|wine_quality| 0.620  | 0.618  |
|superconductivity| 9.523 | 9.443  |
|miami_housing| 84797  | 82022  |
|houses| 0.2199  | 0.2152  |
|healthcare_insurance_expenses| 4415.3  | 4410.1  |
|diamonds| 566.17  | 525.7  |
|concrete_compressive_strength| 4.386  | 4.303  |
|airfoil_self_noise| 1.477  | 1.448  |
|QSAR_fish_toxicity| 0.888  | 0.907  |
|Another-Dataset-on-used-Fiat-500| 753.8  | 750.2  |


|dataset | xgboost_best  | gbt with new hyperparameter |
|-------------| ------------- | ------------- |
|wine_quality| 0.624  | 0.614  |
|superconductivity| 9.553 | 9.444  |
|miami_housing| 83774  | 80987  |
|houses| 0.2199  | 0.2152  |
|healthcare_insurance_expenses| 4649.1  | 4620.3  |
|diamonds| 571.4  | 529.9  |
|concrete_compressive_strength| 4.498 | 4.409  |
|airfoil_self_noise| 1.592  | 1.503  |
|QSAR_fish_toxicity| 0.910  | 0.916  |
|Another-Dataset-on-used-Fiat-500| 740.3  | 728.9  |

**Comment by RAMitchell** [comment]
I really like the idea and I think hyperparameters could definitely be studied more. Some things I would like to see before adding new hyperparameters to xgboost:

- Some kind of theoretical justification - perhaps stronger than because ANOVA does it. I'd also note that xgboost commonly reuses features at multiple depths so depth does not necessarily equal more feature interactions. In general there is certainly a link between tree depth and model complexity. Also note that trees can be sparse, i.e. have high depth but far less than 2^depth leaves.
- Where does the factor of 0.8 come from? Are there other ways to decay (i.e. based on number of leaves?)
- More hyperparameters can be confusing and add too many options for users. Does this just add another dimension that they have to grid search over? Or can it replace other parameters and be simplifying.

**Comment by mgross** [comment]
@RAMitchell 

- I am not sure if a theoretical justification can be made. However, apart from intuitive arguments, I think that pretty strong empirical justifications can be made. In particular, the effect sizes usually shrink as feature interaction depth increases. I performed a mini-study on the 10 datasets chosen above and I think this should generalize pretty well over any sample of real-world datasets (see below). 
  -  On your note that xgboost commonly resuses features at multiple tree depths: One can slightly alter my proposed suggestion by decaying not by tree depth but rather by number of different features used unto the current split (but not sure if this really makes a difference in practice).
- The factor of 0.8 was just for illustration purposes. Over the 10 datasets, the optimal factor or `decay` hyperparameter after grid searching ranged between 0.5 and 0.75. With `decay` = 1.0 exactly the same behaviour as the current xgboost implementation is preserved. There are surely other ways to implement a similar regularization by feature interaction depth or tree-depth (but I am not sure how to base it on number of leaves). Another option would be to e.g. increase `lambda` by a certain percentage or absolute amount depending on tree-depth.
- I agree. The proposed `decay` hyperparameter should be used as a replacement to `max_depth` as tuning both doesn't make sense at all. When tuning `decay`, `max_depth` has just to be set to a fixed and sufficiently high value (12-15), which admittely will slow down fitting somewhat.


----------

For each of the 10 datasets a ridge-regression (using glmnet lib) with all features (standardised to sd=1) up to interaction-depth 4 was performed. Each time, the absolute effect size was averaged by interaction depth. The relative average effect size is visualized here:
<img width="566" height="470" alt="Image" src="https://github.com/user-attachments/assets/3175f28d-0f90-4ccb-8490-4727a9a82426" />
Compared to the main effects (interaction depth = 1), we see (except for one dataset) that the average absolute effect size shrinks with higher interaction depth. This was done using a linear ridge regression model but it should largely transfer to tree models as well. The suggested hyperparameter, tries to acknowledge for this "prior" knowledge on real world datasets.

**Comment by RAMitchell** [comment]
@mgross Sorry for the slow response. I still think this idea is a good one. At the moment I don't think I would be looking to add more hyperparameters to xgboost without serious justification in the form of an academic paper. In the future I would basically be looking at doing a meta study of hyperparameters. I suspect we could effectively do dimension reduction on hyperparameters and that a much smaller number of parameters could explain most of the variance in training outcomes. Something like this could be a candidate in such a study.
